### PR TITLE
feat(stack): Add deprecation date to Stacks

### PR DESCRIPTION
**Pull request by aurelien-reeves-scalingo**
Refs. https://github.com/scalingo/api/issues/1831

- [x] Add a [changelog entry](https://changelog.scalingo.com/)

### CODE REVIEW

**Review by EtienneM — COMMENTED** [review]
What do you think about this solution: https://stackoverflow.com/a/39180230/1351436 to unmarshal the time here? I find it slightly more elegant :)

File: stacks.go [hunk 1]
```diff
@@ -29,10 +44,39 @@ func (c *Client) StacksList(ctx context.Context) ([]Stack, error) {
...
-	return resmap["stacks"], nil
+
+	return jsonStackToStack(resmap["stacks"]), nil
+}
+
+func jsonStackToStack(s []jsonStack) []Stack {
+	var stacks []Stack
+
+	for _, stack := range s {
+		deprecationDate, _ := time.Parse("2006-01-02", stack.DeprecatedAt)
```
Comment (EtienneM) [inline stacks.go:60]: You shoudn't ignore the error btw

**Comment by aurelien-reeves-scalingo** [comment]
> What do you think about this solution: https://stackoverflow.com/a/39180230/1351436 to unmarshal the time here? I find it slightly more elegant :)

I tried that without success. Certainly due to my lack of knowledge of go-lang at the moment.

I would be pleased to pair on that actually :)

**Comment by aurelien-reeves-scalingo** [comment]
I finally managed to refactor it based on your suggest, thanks a lot @EtienneM

File: stacks.go [hunk 2]
```diff
@@ -9,13 +9,18 @@ import (
...
-	Description string    `json:"description"`
-	BaseImage   string    `json:"base_image"`
-	Default     bool      `json:"default"`
+	ID           string          `json:"id"`
+	CreatedAt    time.Time       `json:"created_at"`
+	Name         string          `json:"name"`
+	Description  string          `json:"description"`
+	BaseImage    string          `json:"base_image"`
+	Default      bool            `json:"default"`
+	DeprecatedAt DeprecationDate `json:"deprecated_at"`
```
Comment (EtienneM) [inline stacks.go:23]:
```suggestion
	DeprecatedAt DeprecationDate `json:"deprecated_at,omitempty"`
```

File: stacks_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	},
+}
+
+func TestStackIsDeprecated(t *testing.T) {
+	for message, test := range isDeprecatedCases {
+		t.Run(message, func(t *testing.T) {
+			stack := Stack{DeprecatedAt: DeprecationDate{test.date}}
+			if stack.IsDeprecated() != test.deprecated {
+				t.Errorf("IsDeprecated expected to be %t, got %t", test.deprecated, stack.IsDeprecated())
+			}
```
Comment (EtienneM) [inline stacks_test.go:44]: That's how we usually do:
```suggestion
assert.Equal(t, test.deprecated, stack.IsDeprecated())
```

You can refer to the other tests in the lib to check how this dependency works.

File: stacks_test.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	"test isDeprecated is true when deprecation date is today's date":   {time.Now(), true},
+	"test isDeprecated is true when deprecation date is in the past":    {time.Now().AddDate(0, 0, -1), true},
+	"test isDeprecated is false when deprecation date is in the future": {time.Now().AddDate(0, 0, 1), false},
+}
+
+var stacksListCases = map[string]struct {
+	json          string
+	expectedStack Stack
+}{
+	"test StacksList with stacks with deprecated_at being null": {
```
[the remark this answers is not in the document]
Comment (aurelien-reeves-scalingo) [reply stacks_test.go:28]: This is what it returns on my local environment

image

File: stacks_test.go (hunk 3 above)
Comment (aurelien-reeves-scalingo) [reply stacks_test.go]: That's much better indeed  🤩

**Review by EtienneM — APPROVED** [review]
Niiiiiiiiiiice 🤩

**Pull request merged**